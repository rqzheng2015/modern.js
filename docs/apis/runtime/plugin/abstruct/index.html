<!doctype html>
<html class="docs-version-current" lang="zh-cn" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=0.86,maximum-scale=3,minimum-scale=0.86">
<meta name="generator" content="Docusaurus v2.0.0-beta.6">
<link rel="alternate" type="application/rss+xml" href="/blog/rss.xml" title="Modern.js - 现代 Web 工程体系 Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/blog/atom.xml" title="Modern.js - 现代 Web 工程体系 Blog Atom Feed">

              <script>!function(e,r,n,t,s,o,a,i,c,l,d,f,m,p){o="precollect",a="getAttribute",i="addEventListener",(l=function(e){(d=[].slice.call(arguments)).push(Date.now(),location.href),(e==o?l.p.a:l.q).push(d)}).q=[],l.p={a:[]},e[s]=l,(f=document.createElement("script")).src=n+"?bid=modernjs_dev&globalName="+s,f.crossorigin=n.indexOf("sdk-web")>0?"anonymous":"use-credentials",r.getElementsByTagName("head")[0].appendChild(f),i in e&&(l.pcErr=function(r){r=r||e.event,(m=r.target||r.srcElement)instanceof Element||m instanceof HTMLElement?m[a]("integrity")?e[s](o,"sri",m[a]("href")||m[a]("src")):e[s](o,"st",{tagName:m.tagName,url:m[a]("href")||m[a]("src")}):e[s](o,"err",r.error)},l.pcRej=function(r){r=r||e.event,e[s](o,"err",r.reason||r.detail&&r.detail.reason)},e[i]("error",l.pcErr,!0),e[i]("unhandledrejection",l.pcRej,!0)),"PerformanceLongTaskTiming"in e&&((p=l.pp={entries:[]}).observer=new PerformanceObserver((function(e){p.entries=p.entries.concat(e.getEntries)})),p.observer.observe({entryTypes:["longtask","largest-contentful-paint","layout-shift"]}))}(window,document,"https://lf3-short.bytegoofy.com/slardar/fe/sdk-web/browser.cn.js",0,"Slardar")</script>
              

              <script>window.Slardar("init",{bid:"modernjs_dev"}),window.Slardar("start")</script><title data-react-helmet="true">概览 | Modern.js - 现代 Web 工程体系</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://modernjs.dev//docs/apis/runtime/plugin/abstruct"><meta data-react-helmet="true" name="docusaurus_locale" content="zh-cn"><meta data-react-helmet="true" name="docusaurus_version" content="current"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="概览 | Modern.js - 现代 Web 工程体系"><meta data-react-helmet="true" name="description" content="Modern.js 的基础的插件系统中主要分为三个部分：Hook 模型（Pipeline、Workflow、Waterfall）、 Hook 模型的管理器（Manager），上下文共享机制。"><meta data-react-helmet="true" property="og:description" content="Modern.js 的基础的插件系统中主要分为三个部分：Hook 模型（Pipeline、Workflow、Waterfall）、 Hook 模型的管理器（Manager），上下文共享机制。"><link data-react-helmet="true" rel="shortcut icon" href="/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://modernjs.dev//docs/apis/runtime/plugin/abstruct"><link data-react-helmet="true" rel="alternate" href="https://modernjs.dev//en/docs/apis/runtime/plugin/abstruct" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://modernjs.dev//docs/apis/runtime/plugin/abstruct" hreflang="zh-cn"><link data-react-helmet="true" rel="alternate" href="https://modernjs.dev//docs/apis/runtime/plugin/abstruct" hreflang="x-default"><link rel="stylesheet" href="//lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/css/styles.73bb4626.css">
<link rel="preload" href="//lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/js/runtime~main.d92938dc.js" as="script">
<link rel="preload" href="//lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/js/main.3d6583ba.js" as="script">
</head>
<body itemscope="" itemtype="http://schema.org/Organization">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_V1Ig">跳转到主文档</a></div><nav class="navbar navbar--fixed-top navbarHideable_2J4B"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/"><img src="https://lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/images/images/modernjs-logo.svg" alt="Modernjs Logo" class="themedImage_42gu themedImage--light_-FJF navbar__logo"><img src="https://lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/images/images/modernjs-logo.svg" alt="Modernjs Logo" class="themedImage_42gu themedImage--dark_p0fW navbar__logo"></a><a class="navbar__item navbar__link" href="/docs/start/mobile">开始</a><a class="navbar__item navbar__link" href="/docs/guides/overview">指南</a><a class="navbar__item navbar__link navbar__link--active" href="/docs/apis/overview">API</a><a class="navbar__item navbar__link" href="/blog">博客</a><a class="navbar__item navbar__link" href="/coming-soon">社区</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/modern-js-dev/modern.js" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_91lq react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_P1Jd">🌜</span></div><div class="react-toggle-track-x"><span class="toggle_P1Jd">🌞</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="navbar__search searchBarContainer_0B71"><input placeholder="搜索" aria-label="Search" class="navbar__search-input"><div class="loadingRing_2Ffh searchBarLoadingRing_nyLg"><div></div><div></div><div></div><div></div></div></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_aEDd"><button class="clean-btn backToTopButton_N4sD" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_mm6A"><div class="sidebar_khEP sidebarWithHideableNavbar_j1mR"><a tabindex="-1" class="sidebarLogo_2ggo" href="/"><img src="https://lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/images/images/modernjs-logo.svg" alt="Modernjs Logo" class="themedImage_42gu themedImage--light_-FJF"><img src="https://lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/images/images/modernjs-logo.svg" alt="Modernjs Logo" class="themedImage_42gu themedImage--dark_p0fW"></a><nav class="menu thin-scrollbar menu_2Snv menuWithAnnouncementBar_RH5t"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" href="/docs/apis/overview">概览</a></li><li class="theme-doc-sidebar-item-category menu__list-item"><a class="menu__link menu__link--sublist" href="#">命令</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/commands/overview">概览</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">应用工程方案</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">模块工程方案</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">项目组工程方案</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">工具</a></li></ul></li><li class="theme-doc-sidebar-item-category menu__list-item"><a class="menu__link menu__link--sublist" href="#">文件约定</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/hooks/overview">概览</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">应用工程方案</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">模块工程方案</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">项目组工程方案</a></li></ul></li><li class="theme-doc-sidebar-item-category menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">运行时</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/overview">概览</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/default-alias">默认别名</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/env">环境变量</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">App API</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">BFF Server API</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Model API</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Container API</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Testing API</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Web Server API</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Utility API</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Electron API</a></li><li class="theme-doc-sidebar-item-category menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">插件 API</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/docs/apis/runtime/plugin/abstruct">概览</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/plugin/hook">Hook 模型</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/plugin/manager">Manager</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/plugin/context">上下文共享机制</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/plugin/context-api">Context API</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/runtime/plugin/hook-api">Hook API</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category menu__list-item"><a class="menu__link menu__link--sublist" href="#">配置</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/config/overview">概览</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/config/usage">配置使用</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">source (源文件)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">output (构建产物)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">runtime (运行时)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">server (服务器)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">bff (BFF API)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">electron (Electron)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">dev (开发调试)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">testing (测试)</a></li><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/config/plugins">plugins (框架插件)</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">tools (底层配置)</a></li></ul></li><li class="theme-doc-sidebar-item-category menu__list-item"><a class="menu__link menu__link--sublist" href="#">定制化工程方案</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link menu__list-item"><a class="menu__link" tabindex="0" href="/docs/apis/generator/overview">概览</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">工程方案配置</a></li><li class="theme-doc-sidebar-item-category menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">生成器插件 API</a></li></ul></li></ul></nav></div></aside><main class="docMainContainer_Ip-H"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_JYLG"><div class="docItemContainer_Cd72"><article><div class="theme-doc-markdown markdown"><header><h1>概览</h1></header><p>Modern.js 的基础的插件系统中主要分为三个部分：Hook 模型（Pipeline、Workflow、Waterfall）、 Hook 模型的管理器（Manager），上下文共享机制。</p><p>其中的 Hook 模型是用于管理运行一系列相同模型（形状）函数的管理工具，目前提供了三个大类，7个小类：</p><ul><li>Pipeline<ul><li>Sync</li><li>Async</li></ul></li><li>Waterfall<ul><li>Sync</li><li>Async</li></ul></li><li>Workflow<ul><li>Sync</li><li>Async</li><li>Parallel(Async)</li></ul></li></ul><p>它们之间的区别是他们所管理的函数的运行模式的不同。不同的地方会在下面具体分析介绍。</p><p>Hook 模型的管理器（Manager），顾名思义就是用来管理上面提到的那些 Hook 模型，添加至同一个 Manager 可以使用同一个 Runner 对象（包含所有 Hook 的执行函数）来运行，也共享同一个上下文环境。而 Modern.js 的基础插件系统中的插件就是针对 Manager 来说的，是一个拥有属性、插件 Hook 函数的对象。</p><p>基于上面的工具（Hook 模型 + Manager），在 Modern.js 中主要构建了三套插件模型：CLI、Runtime、Server。其中 CLI 是 Modern.js 中主要的运行流程控制模型，Modern.js 中的各种工程方案（MWA、模块工程方案、Monorepo 工程方案）都是主要通过这一套模型运行的。而 Runtime 则主要负责的是 React 路由组件和元素的处理（Component 和 Element）和渲染（包括服务器端渲染和客户端渲染）。而 Server 则是针对 Server 运行时的这个阶段的生命周期运行和特殊信息收集。其中 Runtime 和 Server 的运行都是在 CLI 中触发的。</p></div></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/docs/apis/runtime/electron/tests/test"><div class="pagination-nav__sublabel">上一篇</div><div class="pagination-nav__label">« Electron Test</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/docs/apis/runtime/plugin/hook"><div class="pagination-nav__sublabel">下一篇</div><div class="pagination-nav__label">Hook 模型 »</div></a></div></nav></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">使用指南</div><ul class="footer__items"><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/docs/start/mobile">快速上手</a></li><li class="footer__item"><a class="footer__link-item" href="/docs/guides/tutorials/overview">实战教程</a></li><li class="footer__item"><a class="footer__link-item" href="/docs/guides/usages/eslint">更多用法</a></li><li class="footer__item"><a class="footer__link-item" href="/docs/guides/features/runtime/code-spliting">专题详解</a></li></ul></ul></div><div class="col footer__col"><div class="footer__title">API 资料</div><ul class="footer__items"><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/docs/apis/commands/overview">命令</a></li><li class="footer__item"><a class="footer__link-item" href="/docs/apis/runtime/overview">运行时</a></li><li class="footer__item"><a class="footer__link-item" href="/docs/apis/hooks/overview">文件约定</a></li><li class="footer__item"><a class="footer__link-item" href="/docs/apis/config/overview">配置选项</a></li></ul></ul></div><div class="col footer__col"><div class="footer__title">关注我们</div><ul class="footer__items"><ul class="footer__items footerItemIcons_L3RU"><a class="footer__item footerItemIcon_9u-Q"><img width="32" height="32" src="https://lf3-static.bytednsdoc.com/obj/eden-cn/aphqeh7uhohpquloj/modern-js/icons/weixin.png"><span class="qrcodeWrap_pGOd" style="display:none"><img width="108" height="108" src="https://lf3-static.bytednsdoc.com/obj/eden-cn/aphqeh7uhohpquloj/modern-js/icons/weixin-qrcode.jpeg" alt="微信"></span></a><a href="https://space.bilibili.com/1195398938" class="footer__item footerItemIcon_9u-Q"><img width="32" height="32" src="https://lf3-static.bytednsdoc.com/obj/eden-cn/aphqeh7uhohpquloj/modern-js/icons/bilibili.png"></a><a class="footer__item footerItemIcon_9u-Q"><img width="32" height="32" src="https://lf3-static.bytednsdoc.com/obj/eden-cn/aphqeh7uhohpquloj/modern-js/icons/feishu.png"><span class="qrcodeWrap_pGOd" style="display:none"><img width="108" height="108" src="https://lf3-static.bytednsdoc.com/obj/eden-cn/aphqeh7uhohpquloj/modern-js/icons/feishu-qrcode.png" alt="飞书"></span></a><a href="https://github.com/modern-js-dev/modern.js" class="footer__item footerItemIcon_9u-Q"><img width="32" height="32" src="https://lf3-static.bytednsdoc.com/obj/eden-cn/aphqeh7uhohpquloj/modern-js/icons/github.png"></a></ul></ul></div></div><div class="footer__bottom text--center"><div class="margin-bottom--sm"><img src="https://lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/images/images/modernjs-logo.svg" alt="ByteDance Open Source Logo" class="themedImage_42gu themedImage--light_-FJF footer__logo"><img src="https://lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/images/images/modernjs-logo.svg" alt="ByteDance Open Source Logo" class="themedImage_42gu themedImage--dark_p0fW footer__logo"></div></div></div></footer></div>
<script src="//lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/js/runtime~main.d92938dc.js"></script>
<script src="//lf-cdn-tos.bytescm.com/obj/static/webinfra/modern-js-website/assets/js/main.3d6583ba.js"></script>
</body>
</html>