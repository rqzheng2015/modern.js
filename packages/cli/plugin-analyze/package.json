{
  "name": "@modern-js/plugin-analyze",
  "description": "The meta-framework suite designed from scratch for frontend-focused modern web development.",
  "homepage": "https://modernjs.dev",
  "bugs": "https://github.com/modern-js-dev/modern.js/issues",
  "repository": "modern-js-dev/modern.js",
  "license": "MIT",
  "keywords": [
    "react",
    "framework",
    "modern",
    "modern.js"
  ],
  "version": "1.0.0",
  "jsnext:source": "./src/index.ts",
  "types": "./dist/types/index.d.ts",
  "main": "./dist/js/node/index.js",
  "module": "./dist/js/treeshaking/index.js",
  "jsnext:modern": "./dist/js/modern/index.js",
  "exports": {
    ".": {
      "node": {
        "import": "./dist/js/modern/index.js",
        "require": "./dist/js/node/index.js"
      },
      "default": "./dist/js/treeshaking/index.js"
    },
    "./cli": "./dist/js/node/index.js"
  },
  "scripts": {
    "prepare": "pnpm build",
    "prepublishOnly": "pnpm build -- --platform",
    "new": "modern new",
    "build": "modern build",
    "test": "modern test --passWithNoTests"
  },
  "dependencies": {
    "@babel/parser": "^7.15.2",
    "@babel/runtime": "^7",
    "@babel/traverse": "^7.15.0",
    "@babel/types": "^7.15.0",
    "@modern-js/plugin": "workspace:^1.0.0",
    "@modern-js/utils": "workspace:^1.0.0",
    "@rollup/pluginutils": "^4.1.1",
    "fs-extra": "^10.0.0",
    "url-join": "^4.0.1"
  },
  "devDependencies": {
    "@modern-js/core": "workspace:^1.0.0",
    "@modern-js/module-tools": "^1.0.0",
    "@modern-js/plugin-testing": "^1.0.0",
    "@modern-js/types": "workspace:^1.0.0",
    "@types/babel__traverse": "^7.14.2",
    "@types/fs-extra": "^9.0.12",
    "@types/jest": "^26",
    "@types/node": "^14",
    "@types/url-join": "^4.0.1",
    "typescript": "^4"
  },
  "peerDependencies": {
    "@modern-js/core": "workspace:^1.0.0"
  },
  "sideEffects": false,
  "modernConfig": {
    "output": {
      "packageMode": "node-js"
    }
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  }
}
