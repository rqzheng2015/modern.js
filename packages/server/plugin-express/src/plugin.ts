import * as path from 'path';
import express, { RequestHandler, Express } from 'express';
import type { Request, Response } from 'express';
import cookieParser from 'cookie-parser';
import { createPlugin } from '@modern-js/server-plugin';
import { requireModule } from '@modern-js/bff-utils';
import { fs, createDebugger } from '@modern-js/utils';
import finalhandler from 'finalhandler';
import { run } from './context';
import registerRoutes from './registerRoutes';

const debug = createDebugger('express');

interface FrameConfig {
  middleware: (RequestHandler | string)[];
}

export type Mode = 'function' | 'framework';

const findAppModule = async (apiDir: string) => {
  const exts = ['.ts', '.js'];
  const paths = exts.map(ext => path.resolve(apiDir, `app${ext}`));

  for (const filename of paths) {
    if (await fs.pathExists(filename)) {
      return requireModule(filename);
    }
  }

  return null;
};

const initMiddlewares = (
  middleware: (RequestHandler | string)[],
  app: Express,
) => {
  middleware.forEach(middlewareItem => {
    const middlewareFunc =
      typeof middlewareItem === 'string'
        ? requireModule(middlewareItem)
        : middlewareItem;
    app.use(middlewareFunc);
  });
};

const useRun = (app: Express) => {
  app.use((req, res, next) => {
    run({ req, res }, next);
  });
};

const initApp = (app: express.Express) => {
  app.use(cookieParser());
  app.use(express.text());
  app.use(express.urlencoded({ extended: true }));
  app.use(express.json());
  return app;
};

export default createPlugin(
  () => ({
    // eslint-disable-next-line max-statements
    async prepareApiServer({ pwd, mode, config }) {
      let app: Express;
      const apiDir = path.join(pwd, './api');
      if (mode === 'framework') {
        app = await findAppModule(apiDir);

        if (!app || !app.use) {
          console.warn('There is not api/app.ts.');
          app = express();
        }
        initApp(app);

        if (config) {
          const { middleware } = config as FrameConfig;
          initMiddlewares(middleware, app);
        }
        // eslint-disable-next-line react-hooks/rules-of-hooks
        useRun(app);

        registerRoutes(app);
      } else if (mode === 'function') {
        app = express();
        initApp(app);

        if (config) {
          const { middleware } = config as FrameConfig;
          initMiddlewares(middleware, app);
        }

        // eslint-disable-next-line react-hooks/rules-of-hooks
        useRun(app);

        registerRoutes(app);
      } else {
        throw new Error(`mode must be function or framework`);
      }

      return (req, res) =>
        new Promise((resolve, reject) => {
          const handler = (err: any) => {
            if (err) {
              return reject(err);
            }
            // finalhanlder will trigger 'finish' event
            return finalhandler(req, res, {})(null);
            // return resolve();
          };

          res.on('finish', (err: Error) => {
            if (err) {
              return reject(err);
            }
            return resolve();
          });
          return app(req as Request, res as Response, handler);
        });
    },

    prepareWebServer({ config }) {
      const app = express();
      initApp(app);
      if (config) {
        const { middleware } = config as FrameConfig;
        debug('web middleware', middleware);
        initMiddlewares(middleware, app);
      }

      return (req, res) =>
        new Promise((resolve, reject) => {
          const handler = (err: string) => {
            if (err) {
              return reject(err);
            }
            if (res.headersSent && res.statusCode !== 200) {
              finalhandler(req, res, {})(null);
            }
            return resolve();
          };

          // when user call res.send
          res.on('finish', (err: Error) => {
            if (err) {
              return reject(err);
            }
            return resolve();
          });
          return app(req as Request, res as Response, handler);
        });
    },
  }),
  {
    name: '@modern-js/plugin-express',
    pre: ['@modern-js/plugin-bff'],
  },
) as any;
