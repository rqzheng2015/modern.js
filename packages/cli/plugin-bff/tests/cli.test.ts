import path from 'path';
import { manager, AppContext } from '@modern-js/core';
import { modifyServerRoutes } from '@modern-js/plugin-analyze';
import Chain from 'webpack-chain';
import plugin from '../src/cli';

const root = path.normalize(path.resolve(__dirname, '../../../../'));
expect.addSnapshotSerializer({
  test: val =>
    typeof val === 'string' &&
    (val.includes('modern.js') ||
      val.includes('node_modules') ||
      val.includes(root)),
  print: val =>
    // eslint-disable-next-line no-nested-ternary
    typeof val === 'string'
      ? // eslint-disable-next-line no-nested-ternary
        val.includes('node_modules')
        ? `"${val.replace(/.+node_modules/, '')}"`
        : val.includes('modern.js')
        ? `"${val.replace(/.+modern\.js/, '')}"`
        : `"${val.replace(root, '')}"`
      : (val as string),
});

describe('bff cli plugin', () => {
  it('schema', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const result = await runner.validateSchema();

    expect(result).toMatchSnapshot();
  });

  it('config', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const result = await runner.config();

    expect(result).toMatchSnapshot();
  });

  it('config webpack chain', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const [{ tools }]: any = await runner.config();
    const chain = new Chain();
    manager.run(() => {
      AppContext.set({
        appDirectory: './fixtures/function',
        port: 3000,
      } as any);
    });
    manager.run(() => tools.webpack({}, { chain }));

    expect(chain.toConfig()).toMatchObject({
      module: {
        rules: [
          {
            oneOf: [
              {
                test: /.\/fixtures\/function\/api\/\.*(\.[tj]s)$/,
                use: [
                  {
                    loader: require
                      .resolve('../src/loader.ts')
                      .replace(/\\/g, '/'),
                    options: {
                      apiDir: path.resolve('./fixtures/function/api'),
                      fetcher: undefined,
                      port: 3000,
                      prefix: '/api',
                      target: undefined,
                    },
                  },
                ],
              },
            ],
          },
        ],
      },
      resolve: {
        alias: {
          '@api': path.resolve('./fixtures/function/api'),
        },
      },
    });
  });

  it('server routes', async () => {
    const main = manager.clone().usePlugin(plugin);
    main.registe({ modifyServerRoutes });
    const runner = await main.init();
    const result = await runner.modifyServerRoutes({ routes: [] });

    expect(result).toMatchSnapshot();
  });
});
