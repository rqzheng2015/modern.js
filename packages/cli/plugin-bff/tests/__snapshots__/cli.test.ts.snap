// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bff cli plugin config 1`] = `
Array [
  Object {
    "source": Object {
      "moduleScopes": Array [
        "./api",
      ],
    },
    "tools": Object {
      "webpack": [Function],
    },
  },
]
`;

exports[`bff cli plugin schema 1`] = `
Array [
  Array [
    Object {
      "schema": Object {
        "properties": Object {
          "fetcher": Object {
            "type": "string",
          },
          "prefix": Object {
            "type": "string",
          },
          "proxy": Object {
            "type": "object",
          },
          "requestCreater": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "target": "bff",
    },
  ],
]
`;

exports[`bff cli plugin server routes 1`] = `
Object {
  "routes": Array [
    Object {
      "entryPath": "",
      "isApi": true,
      "isSPA": false,
      "isSSR": false,
      "urlPath": "/api",
    },
  ],
}
`;
